File: frontend/src/api/emailOutbox.ts
import { api } from './index'

export type EmailOutboxStatus =
  | 'PENDING'
  | 'SENDING'
  | 'SENT'
  | 'FAILED'
  | 'DLQ'
  | 'CANCELLED'

export interface EmailOutboxItem {
  id: string
  kind: string
  to_email: string
  subject: string
  status: EmailOutboxStatus
  attempts: number
  max_attempts: number
  next_attempt_at: string | null
  last_error: string | null
  last_error_type: string | null
  last_error_class: string | null
  related_resource_type: string | null
  related_resource_id: string | null
  created_at: string | null
  updated_at: string | null
  sent_at: string | null
}

export interface EmailOutboxDetail extends EmailOutboxItem {
  body_html: string
  body_text: string | null
  payload: Record<string, unknown>
}

export interface EmailOutboxListOut {
  items: EmailOutboxItem[]
  total: number
  limit: number
  offset: number
  counts_30d: Record<string, number>
}

export interface EmailOutboxFilters {
  status?: EmailOutboxStatus | ''
  kind?: string
  to_email?: string
  q?: string
  date_from?: string
  date_to?: string
  limit?: number
  offset?: number
}

export function fetchEmailOutbox(filters: EmailOutboxFilters = {}) {
  const query: Record<string, string | number> = {}
  for (const [k, v] of Object.entries(filters)) {
    if (v !== undefined && v !== null && v !== '') query[k] = v as string | number
  }
  return api<EmailOutboxListOut>('/admin/email-outbox', { query })
}

export function fetchEmailOutboxItem(id: string) {
  return api<EmailOutboxDetail>(`/admin/email-outbox/${id}`)
}

export function retryEmailOutboxItem(id: string, resetAttempts = true) {
  return api<{ detail: string }>(
    `/admin/email-outbox/${id}/retry?reset_attempts=${resetAttempts}`,
    { method: 'POST' },
  )
}

export function cancelEmailOutboxItem(id: string) {
  return api<{ detail: string }>(`/admin/email-outbox/${id}/cancel`, { method: 'POST' })
}

export function fetchEmailOutboxStats() {
  return api<{ counts: Record<string, number>; oldest_pending_at: string | null }>(
    '/admin/email-outbox/_/stats',
  )
}

File: backend/app/models/email_outbox.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import CheckConstraint, DateTime, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class EmailOutbox(Base):
    __tablename__ = "email_outbox"
    __table_args__ = (
        CheckConstraint(
            "status IN ('PENDING','SENDING','SENT','FAILED','DLQ','CANCELLED')",
            name="ck_email_outbox_status",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    kind: Mapped[str] = mapped_column(String(64), nullable=False)
    to_email: Mapped[str] = mapped_column(String(320), nullable=False)
    subject: Mapped[str] = mapped_column(String(998), nullable=False)
    body_html: Mapped[str] = mapped_column(Text, nullable=False, server_default=text("''"))
    body_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    payload: Mapped[dict[str, Any]] = mapped_column(
        JSONB, nullable=False, server_default=text("'{}'::jsonb"), default=dict
    )
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text("'PENDING'")
    )
    attempts: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("0"))
    max_attempts: Mapped[int] = mapped_column(Integer, nullable=False, server_default=text("6"))
    next_attempt_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("NOW()")
    )
    last_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    last_error_type: Mapped[str | None] = mapped_column(String(128), nullable=True)
    last_error_class: Mapped[str | None] = mapped_column(String(16), nullable=True)
    related_resource_type: Mapped[str | None] = mapped_column(String(64), nullable=True)
    related_resource_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_by_user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("NOW()")
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("NOW()")
    )
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
